File: algorithmic-toolbox/week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>

using std::vector;
using std::cin;
using std::cout;
using std::max;

long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = max(max_product,
                (long long)numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

long long MaxPairwiseProductFast(const vector<int>& numbers) {
    int n = numbers.size();

    int max_index1 = -1;
    for (int first = 0; first < n; ++first) {
        if((max_index1 == -1) || (numbers[first] > numbers[max_index1])) {
            max_index1 = first;
        }
    }

    int max_index2 = -1;
    for (int second = 0; second < n; ++second) {
        if ((second != max_index1) && ((max_index2 == -1) || 
            (numbers[second] > numbers[max_index2]))) {
            max_index2 = second;
        }
    }

    return ((long long)(numbers[max_index1])) * numbers[max_index2];
}

int main() {
    // Stress Test
    /*while (true) {
        int n = rand() % 10 + 2;
        cout << n << "\n";
        vector<int> a;
        for (int i = 0; i < n; ++i) {
            a.push_back(rand() % 10000);
        }
        for (int j = 0; j < n; ++j) {
            cout << a[j] << ' ';
        }
        cout << "\n";
        long long result1 = MaxPairwiseProduct(a);
        long long result2 = MaxPairwiseProductFast(a);
        if (result1 != result2) {
            cout << "Wrong! " << result1 << ' ' << result2 << "\n";
            break;
        }
        else {
            cout << "OK\n";
        }
    }*/
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProductFast(numbers) << "\n";
    return 0;
}

File: algorithmic-toolbox/week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.py
# python3


def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product


def max_pairwise_product_fast(numbers):
    n = len(numbers)
    
    max_index1 = -1
    for first in range(n):
        if max_index1 == -1 or (numbers[first] > numbers[max_index1]):
            max_index1 = first

    max_index2 = -1
    for second in range(n):
        if (not second == max_index1) and ((max_index2 == -1) 
            or (numbers[second] > numbers[max_index2])):
            max_index2 = second
    return numbers[max_index1] * numbers[max_index2]


if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product_fast(input_numbers))
